File: macos/Packages/CouchTourKit/Sources/CouchTourKit/ArtistAbbreviations.swift
import Foundation

/// Artist abbreviations for narrow desktop columns and compact badges.
/// Source: etree.org BandAbbreviations wiki, upper case for true acronyms (DMB, SCI, GD),
/// lower case for word nicknames (mule, pgroove). Fall back to name when not listed.
public enum ArtistAbbreviations {
    public static let map: [String: String] = [
        "Allman Brothers Band": "ABB",
        "Bela Fleck & the Flecktones": "BFFT",
        "Ben Harper and the Innocent Criminals": "BHIC",
        "Blues Traveler": "BT",
        "Brothers Past": "BP",
        "Dark Star Orchestra": "DSO",
        "Dave Matthews Band": "DMB",
        "David Nelson Band": "DNB",
        "Deep Banana Blackout": "DBB",
        "Derek Trucks Band": "DTB",
        "Dirty Dozen Brass Band": "DDBB",
        "Disco Biscuits": "DB",
        "Donna the Buffalo": "DTB",
        "Drive-By Truckers": "DBT",
        "Ekoostik Hookah": "EH",
        "God Street Wine": "GSW",
        "Gov't Mule": "mule",
        "Grateful Dead": "GD",
        "GreyBoy AllStars": "GBA",
        "Jazz Mandolin Project": "JMP",
        "Jerry Garcia Band": "JGB",
        "Jerry Joseph & The Jackmormons": "JJJ",
        "Jimi Hendrix Experience": "JHE",
        "John Mayer Trio": "JM3",
        "Karl Denson's Tiny Universe": "KDTU",
        "Keller Williams": "KW",
        "Leftover Salmon": "LS",
        "Les Claypool's Fearless Flying Frog Brigade": "FFFB",
        "Little Feat": "LF",
        "Medeski Martin & Wood": "MMW",
        "Michael Franti & Spearhead": "MFS",
        "North Mississippi Allstars": "NMAS",
        "Pat Metheny Group": "PMG",
        "Pat McGee Band": "PMB",
        "Pearl Jam": "PJ",
        "Phil Lesh & Friends": "phil",
        "Phish": "PH",
        "Railroad Earth": "RRE",
        "Ratdog": "ratdog",
        "Robert Randolph & the Family Band": "RRFB",
        "Rusted Root": "RR",
        "Sound Tribe Sector 9": "STS9",
        "Sound Tribe Sector9": "STS9",
        "Steve Kimock Band": "SKB",
        "String Cheese Incident": "SCI",
        "Tea Leaf Green": "TLG",
        "The Big Wu": "wu",
        "The Dead": "dead",
        "The New Deal": "TND",
        "The Slip": "slip",
        "Trey Anastasio Band": "TAB",
        "Umphrey's McGee": "UM",
        "Widespread Panic": "WSP",
        "Yonder Mountain String Band": "YMSB",
        "Perpetual Groove": "pgroove",
        "Goose": "goose",
        "moe.": "moe.",
    ]

    /// Returns abbreviation if fits is false, full name otherwise.
    /// Preserves "moe." lowercase with period (uat-005).
    public static func label(for name: String, fits: Bool = true) -> String {
        if name.lowercased() == "moe." || name.lowercased() == "moe" {
            return "moe."
        }
        if fits { return name }
        return map[name] ?? name
    }
}

File: app/src/main/java/dev/mike/couchtour/ArtistAbbreviations.kt
package dev.mike.couchtour

/**
 * Artist display names for narrow columns (e.g. Home date-first rows, Library rows).
 * Sourced from etree.org BandAbbreviations wiki (the codes tapers already use in filenames).
 * Acronyms are uppercase (GD, DMB, SCI, UM); nicknames are lowercase (mule, pgroove, goose).
 */
object ArtistAbbreviations {

    val ABBREVIATIONS: Map<String, String> = mapOf(
        "Allman Brothers Band" to "ABB",
        "Bela Fleck & the Flecktones" to "BFFT",
        "Ben Harper and the Innocent Criminals" to "BHIC",
        "Blues Traveler" to "BT",
        "Brothers Past" to "BP",
        "Dark Star Orchestra" to "DSO",
        "Dave Matthews Band" to "DMB",
        "David Nelson Band" to "DNB",
        "Deep Banana Blackout" to "DBB",
        "Derek Trucks Band" to "DTB",
        "Dirty Dozen Brass Band" to "DDBB",
        "Disco Biscuits" to "DB",
        "Donna the Buffalo" to "DTB",
        "Drive-By Truckers" to "DBT",
        "Ekoostik Hookah" to "EH",
        "God Street Wine" to "GSW",
        "Gov't Mule" to "mule",
        "Grateful Dead" to "GD",
        "GreyBoy AllStars" to "GBA",
        "Jazz Mandolin Project" to "JMP",
        "Jerry Garcia Band" to "JGB",
        "Jerry Joseph & The Jackmormons" to "JJJ",
        "Jimi Hendrix Experience" to "JHE",
        "John Mayer Trio" to "JM3",
        "Karl Denson's Tiny Universe" to "KDTU",
        "Keller Williams" to "KW",
        "Leftover Salmon" to "LS",
        "Les Claypool's Fearless Flying Frog Brigade" to "FFFB",
        "Little Feat" to "LF",
        "Medeski Martin & Wood" to "MMW",
        "Michael Franti & Spearhead" to "MFS",
        "North Mississippi Allstars" to "NMAS",
        "Pat Metheny Group" to "PMG",
        "Pat McGee Band" to "PMB",
        "Pearl Jam" to "PJ",
        "Phil Lesh & Friends" to "phil",
        "Phish" to "PH",
        "Railroad Earth" to "RRE",
        "Ratdog" to "ratdog",
        "Robert Randolph & the Family Band" to "RRFB",
        "Rusted Root" to "RR",
        "Sound Tribe Sector 9" to "STS9",
        "Steve Kimock Band" to "SKB",
        "String Cheese Incident" to "SCI",
        "Tea Leaf Green" to "TLG",
        "The Big Wu" to "wu",
        "The Dead" to "dead",
        "The New Deal" to "TND",
        "The Slip" to "slip",
        "Trey Anastasio Band" to "TAB",
        "Umphrey's McGee" to "UM",
        "Widespread Panic" to "WSP",
        "Yonder Mountain String Band" to "YMSB",
        "Perpetual Groove" to "pgroove",
        "Goose" to "goose",
    )

    /**
     * Formats the artist name, substituting known abbreviations when it does not fit
     * or when abbreviations are preferred for compact column layouts.
     * Also respects uat-005: 'moe.' is always rendered in lowercase with a trailing period.
     */
    fun artistLabel(name: String, fits: Boolean = true): String {
        if (name.equals("moe.", ignoreCase = true) || name.equals("moe", ignoreCase = true)) {
            return "moe."
        }
        if (fits) return name
        return ABBREVIATIONS[name] ?: name
    }
}
